File: server/consul.rb
require "socket"
require 'rest-client'


consul_url = ENV['CONSUL_URL'] || "http://localhost:8500"

def get_ip
    ip = Socket.ip_address_list.detect(&:ipv4_private?).ip_address
end

def register_service(consul_url)
    headers = { 'Content-Type': 'application/json' }
  
    random_id = SecureRandom.hex(5)
  
    service_name = "backend"
    ip = get_ip

    payload = {
      "ID": service_name + "-" + random_id,
      "Name": service_name,
      "Tags": ["api"],
      "Address":  ip,
      "Port": 4567,
      "Check": {
        "DeregisterCriticalServiceAfter": "90m",
        "HTTP": "http://#{ip}:4567/health",
        "Interval": "60s",
        "Timeout": "60s"
      }
    }
    RestClient.put(consul_url, payload.to_json, headers)
end

register_service(consul_url  + '/v1/agent/service/register')

File: frontend-project/src/consul.ts
const registerService = async () => {
  const url = 'http://localhost:8500/v1/agent/service/register';
  const data = {
    ID: 'frontend2',
    Name: 'frontend2',
    Address: '127.0.0.1',
    Port: 5173,
  };

  const options = {
    method: 'PUT',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(data),
  };

  try {
    const response = await fetch(url, options);

    if (response.ok) {
      const responseData = await response.json();
      console.log('Servicio registrado en Consul:', responseData);
    } else {
      console.error('Error al registrar el servicio en Consul:', response.statusText);
    }
  } catch (error) {
    console.error('Error desconocido:', error);
  }
};

registerService();
